src/sentry/rules/history/preview.py

```diff
@@ -201,5 +273,89 @@ def get_events(
+) -> GroupActivityMap:
+    """
+    Applies frequency conditions to issue state activity.
+    """
+    conditions = []
+    for condition_data in frequency_conditions:
+        condition_cls = rules.get(condition_data["id"])
+        if condition_cls is None:
+            raise PreviewException
+        conditions.append(condition_cls(project, data=condition_data))
+
+    filtered_activity = defaultdict(list)
+    for group, activities in group_activity.items():
+        # TODO: only EventFrequencyCondition is supported right now, so we can reuse the same query
+        buckets = get_frequency_buckets(project, start, end, group)
+        pass_count = [0] * len(activities)
+        for condition in conditions:
+            for i, activity in enumerate(activities):
+                try:
+                    if condition.passes_activity_frequency(activity, buckets):
+                        pass_count[i] += 1
+                except NotImplementedError:
+                    raise PreviewException
+
+            for i in range(len(activities)):
+                if (
+                    pass_count[i]
+                    and condition_match == "any"
+                    or pass_count[i] == len(conditions)
+                    and condition_match == "all"
+                ):
+                    filtered_activity[group].append(activities[i])
+
+    return filtered_activity
+
+
+def get_frequency_buckets(
+    project: Project, start: datetime, end: datetime, group_id: str
+) -> Dict[datetime, int]:
+    """
+    Puts the events of a group into buckets, and returns the bucket counts.
+    The returned list contains the prefix sum of the buckets to support efficient range queries by frequency conditions.
+    """
+    # TODO: support counting of other issue types and fields (# of unique users, ...)
+    bucket_counts = raw_query(
+        dataset=Dataset.Events,
+        start=start,
+        end=end,
+        filter_keys={"project_id": [project.id]},
+        conditions=[("group_id", "=", group_id)],
+        aggregations=[
+            ("toStartOfFiveMinute", "timestamp", "roundedTime"),
+            ("count", "roundedTime", "bucketCount"),
+        ],
+        groupby=["roundedTime"],
+        orderby=["-roundedTime"],
+        selected_columns=["roundedTime", "bucketCount"],
+        limit=PREVIEW_TIME_RANGE // FREQUENCY_CONDITION_BUCKET_SIZE + 1,  # at most ~4k
+    ).get("data", [])
+    for bucket in bucket_counts:
+        bucket["roundedTime"] = parse_snuba_datetime(bucket["roundedTime"])
+
+    rounded_time = round_to_five_minute(start)
+    rounded_end = round_to_five_minute(end)
+    cumulative_sum = 0
+    buckets = {}
+    # the query result only contains buckets that have a positive count
+    # here we fill in the empty buckets and accumulate the sum
+    while rounded_time <= rounded_end:
+        if bucket_counts and bucket_counts[-1]["roundedTime"] == rounded_time:
+            cumulative_sum += bucket_counts.pop()["bucketCount"]
```

Comment:
don't really love the rounding here and timestamp comparison here. is there any way we can keep the timestamps/buckets as normal and then use the timestamps later implicitly? happy to whiteboard this with you